Source/OrchBase/Placement/Algorithms/SimulatedAnnealing.cpp

```diff
@@ -0,0 +1,436 @@
+     * rounding. */
+    float finalFitness = placer->compute_fitness(task);
+    time = placer->timestamp();
+    result.endTime = time;
+    placer->populate_result_structures(&result, task, finalFitness);
+    fprintf(log, "[I] Final fitness from deltas: %f, Iteration count: %d, "
+            "actual final fitness: %f\n",
+            fitness, iteration, finalFitness);
+    fprintf(log, "[I] Placement complete at %s.\n", time.c_str());
+    fclose(log);
+    fclose(data);
+    return fitness;
+}
+
+/* Returns true if the termination condition holds, and false otherwise. It's
+ * pretty simple - for now we're just counting iterations - but this probably
+ * will change as we become more curious. */
+bool SimulatedAnnealing::is_finished(){return iteration >= ITERATION_MAX;}
+
+/* Performs the selection operation for simulated annealing. Arguments
+ *
+ * - task: Holds the graph from which devices are selected.
+ *
+ * - device: Pointer to a device pointer to select for the atomic
+ *   transformation.
+ *
+ * - thread: Pointer to a thread pointer to either move the aforementioned
+ *   device to, or that holds a device to swap with.
+ *
+ * - swapDevice: If the operation is a swap operation, points to the device to
+ *   swap with. If not, is set to PNULL. */
+void SimulatedAnnealing::select(P_task* task, P_device** device,
+                                P_thread** thread, P_device** swapDevice)
+{
+    /* Sanity. */
+    *device = PNULL;
+    *thread = PNULL;
+    *swapDevice = PNULL;
+
+    /* Choose a non-supervisor device. Note that this will loop infinitely if
+     * there are only supervisor devices in this task (but I'm assuming nobody
+     * is going to do that... */
+    do
+    {
+        unsigned nodeKey;  /* Unused */
+        task->pD->G.RandomNode(nodeKey, *device);
+    } while (!((*device)->pP_devtyp->pOnRTS));
+
+    /* Choose a (valid) core. If there are no valid cores, set thread to NULL
+     * and leave. */
+    P_core* core;
+    P_devtyp* deviceType = (*device)->pP_devtyp;
+    std::map<P_devtyp*, std::set<P_core*>>::iterator bigScaryMapIterator;
+    bigScaryMapIterator = validCoresForDeviceType.find(deviceType);
+    if (bigScaryMapIterator->second.empty()) return;
+
+    /* Note that we don't need to check for "unfound" device types, because
+     * validCoresForDeviceType has been populated with all non-supervisor
+     * device types earlier. */
+    std::set<P_core*>::iterator coreIterator;
+    coreIterator = bigScaryMapIterator->second.begin();
+    std::advance(coreIterator, rand() % bigScaryMapIterator->second.size());
+    core = *coreIterator;
+
+    /* Choose a thread on that core. */
+    std::map<AddressComponent, P_thread*>::iterator threadIterator;
+    threadIterator = core->P_threadm.begin();
+    std::advance(threadIterator, rand() % core->P_threadm.size());
+    *thread = threadIterator->second;
+
+    /* <!> NB: Disabling swap operations for now, to make debugging simpler. */
```

Comment:
is this still a valid comment?

Reply:
Yes. Move is quite enough to be getting on with. The only feature swapping supports is if an application fits in the hardware exactly, as the space of "placement mappings" can be completely traversed by move operations.